vsomeip-sys/src/safe_glue.rs

```diff
@@ -0,0 +1,455 @@
+/********************************************************************************
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use crate::cxx_bridge::handler_registration::{
+    offer_single_event, register_availability_handler_fn_ptr, register_message_handler_fn_ptr,
+    register_subscription_status_handler_fn_ptr, request_single_event,
+};
+use crate::extern_callback_wrappers::{
+    AvailabilityHandlerFnPtr, MessageHandlerFnPtr, SubscriptionStatusHandlerFnPtr,
+};
+use crate::ffi::glue::{get_payload_raw, set_payload_raw};
+use crate::glue::upcast;
+use crate::glue::{ApplicationWrapper, MessageWrapper, PayloadWrapper, RuntimeWrapper};
+use crate::unsafe_fns::create_payload_wrapper;
+use crate::vsomeip::message_base;
+use crate::vsomeip::{application, message, payload, runtime};
+use cxx::UniquePtr;
+use log::{error, trace};
+use std::pin::Pin;
+use std::slice;
+
+/// Gets a `Pin<&mut runtime>` from a [RuntimeWrapper]
+///
+/// # Rationale
+///
+/// In order to use the wrapped methods on a [runtime], we must have a `Pin<&mut runtime>`
+///
+/// Because vsomeip makes heavy use of std::shared_ptr and Rust's ownership and borrowing model
+/// doesn't work well together with it, in Rust code use a UniquePtr<[RuntimeWrapper]>
+///
+/// Since we use a UniquePtr<[RuntimeWrapper]>, we then need a way to drill down and extract
+/// the `Pin<&mut runtime>`.
+///
+/// # TODO
+///
+/// Add some runtime safety checks on the pointer
+pub fn get_pinned_runtime(wrapper: &RuntimeWrapper) -> Pin<&mut runtime> {
```

Comment:
can these explicit functions maybe replaced with some clever implementation of `Deref` instead?

Reply:
I guess this is not allowed since the borrow checker will not allow for pinned mutable borrows to be returned from Deref.

I spent some time on this today, but didn't see a way around the issue.

Darn. I'm going to call this one unsolvable for the moment :(

Reply:
I updated this to clarify and hopefully simplify the API.